driver-scala/src/main/scala/org/mongodb/scala/internal/ZipObservable.scala

```diff
@@ -56,38 +60,61 @@ private[scala] case class ZipObservable[T, U](
         }
 
         override def onSubscribe(subscription: Subscription): Unit = {
-          if (firstSub) {
-            observable1Subscription = Some(subscription)
+          if (isLeftSub) {
+            leftSubscription = Some(subscription)
           } else {
-            observable2Subscription = Some(subscription)
+            rightSubscription = Some(subscription)
           }
 
-          if (observable1Subscription.nonEmpty && observable2Subscription.nonEmpty) {
+          if (leftSubscription.nonEmpty && rightSubscription.nonEmpty) {
             observer.onSubscribe(jointSubscription)
           }
         }
 
         override def onComplete(): Unit = {
-          if (!firstSub) {
-            terminated = true
-            observer.onComplete()
-          }
+          markCompleted(isLeftSub)
+          processNext(observer)
         }
 
         override def onNext(tResult: A): Unit = {
+          if (isLeftSub) leftCounter.incrementAndGet() else rightCounter.incrementAndGet()
           counter += 1
           queue.add((counter, tResult))
-          if (!firstSub) processNext(observer)
+          processNext(observer)
         }
       }
     }
 
-    private def processNext(observer: Observer[_ >: (T, U)]): Unit = {
-      (thisQueue.peek, thatQueue.peek) match {
-        case ((k1: Long, _), (k2: Long, _)) if k1 == k2 => observer.onNext((thisQueue.poll()._2, thatQueue.poll()._2))
+    private def markCompleted(isLeftSub: Boolean): Unit = synchronized {
+      if (isLeftSub) {
+        completedLeft = true
+      } else {
+        completedRight = true
+      }
+    }
+
+    private def completed(): Unit = synchronized {
+      if (!terminated) {
+        terminated = true
+        leftSubscription.foreach(_.unsubscribe())
+        rightSubscription.foreach(_.unsubscribe())
+        observer.onComplete()
+      }
+    }
+
+    private def processNext(observer: Observer[_ >: (L, R)]): Unit = synchronized {
+      (leftQueue.peek, rightQueue.peek) match {
+        case ((k1: Long, _), (k2: Long, _)) if k1 == k2 =>
+          observer.onNext((leftQueue.poll()._2, rightQueue.poll()._2))
+          processNext(observer)
         case _ =>
-          if (!terminated && !jointSubscription.isUnsubscribed) jointSubscription.request(1) // Uneven queues request more data
-        // from downstream so to honor the original request for data.
+          if (!terminated && !jointSubscription.isUnsubscribed) {
+            if (completedLeft && rightCounter.get() >= leftCounter.get()) {
```

Comment:
Can you explain the second part of the conditional?  I'm not grokking it... I was expecting just:

     if (completedLeft || completedRight) {
         completed();
     }

Is there a test for the second part of the condition?

Reply:
We can't complete when either left or right completes as the other observable may be suffering from latency and be slower at producing data (the bug this fixes).

So we have two observables (left, right) each potentially producing data at different rates.  
We need to complete the zip of the two observables when either:

1) The left observable has completed and the right observable has produced the same amount or more data than the left.
2) The right observable has completed and the left observable has produced the same amount or more data than the right.

The empty zippedObservablesWithEmptyObservable test cases test this.

Reply:
To help visualize the right observable produces data quicker than the left and finishes first.

![zipMarbles](https://user-images.githubusercontent.com/420/79316510-5029da80-7efc-11ea-9ed8-6a3ea7aa2392.png)

Reply:
Thanks